supreme/src/commonMain/kotlin/at/asitplus/signum/supreme/symmetric/decrypt.kt

```diff
@@ -0,0 +1,206 @@
+import at.asitplus.signum.indispensable.SecretExposure
+import at.asitplus.signum.indispensable.mac.MessageAuthenticationCode
+import at.asitplus.signum.indispensable.symmetric.*
+import at.asitplus.signum.indispensable.symmetric.AuthCapability.Authenticated
+import at.asitplus.signum.indispensable.symmetric.SymmetricEncryptionAlgorithm.AES
+import kotlin.jvm.JvmName
+
+
+/**
+ * Attempts to decrypt this ciphertext (which may also hold an IV/nonce, and in case of an authenticated ciphertext, authenticated data and auth tag) using the provided [key].
+ * This is the generic, untyped decryption function for convenience.
+ * **Compared to its narrower-typed cousins is possible to mismatch the characteristics of
+ * [key] and [SealedBox].**
+ */
+@JvmName("decryptGeneric")
+suspend fun SealedBox<*, *, *>.decrypt(key: SymmetricKey<*, *, *>): KmmResult<ByteArray> = catching {
+    require(algorithm == key.algorithm) { "Algorithm mismatch! expected: $algorithm, actual: ${key.algorithm}" }
+    @Suppress("UNCHECKED_CAST")
+    when (algorithm.authCapability) {
+        is Authenticated.Integrated -> (this as SealedBox<Authenticated.Integrated, *, KeyType.Integrated>).decryptInternal(
+            @OptIn(SecretExposure::class) (key as SymmetricKey.Integrated).secretKey.getOrThrow(),
+            byteArrayOf()
+        )
+
+        is Authenticated.WithDedicatedMac<*, *> -> {
+            key as SymmetricKey.WithDedicatedMac
+            (this as SealedBox<Authenticated.WithDedicatedMac<*, *>, *, KeyType.WithDedicatedMacKey>).decryptInternal(
+                @OptIn(SecretExposure::class) key.encryptionKey.getOrThrow(),
+                @OptIn(SecretExposure::class) key.macKey.getOrThrow(),
+                byteArrayOf()
+            )
+        }
+
+        is AuthCapability.Unauthenticated -> (this as SealedBox<AuthCapability.Unauthenticated, *, KeyType.Integrated>).decryptInternal(
+            @OptIn(SecretExposure::class) (key as SymmetricKey.Integrated).secretKey.getOrThrow()
+        )
+    }
+}
+
+
+//required because we don't store MAC info all the way
+/**
+ * Attempts to decrypt this ciphertext (which may also hold an IV/nonce) using the provided [key].
+ * Do provide [authenticatedData] if required, or else decryption will fail!
+ * This is the generic, untyped decryption function for convenience.
+ * **Compared to its narrower-typed cousins is possible to mismatch the characteristics of
+ * [key] and [SealedBox].**
+ */
+@JvmName("decryptAuthenticatedIntegrated")
+suspend fun <I : NonceTrait, M : MessageAuthenticationCode> SealedBox<AuthCapability.Authenticated.WithDedicatedMac<M, I>, I, KeyType.WithDedicatedMacKey>.decrypt(
+    key: SymmetricKey.WithDedicatedMac<*>,
+    authenticatedData: ByteArray = byteArrayOf()
+) = catching {
+    @Suppress("UNCHECKED_CAST")
+    (this as SealedBox<Authenticated.WithDedicatedMac<*, *>, *, KeyType.WithDedicatedMacKey>).decryptInternal(
+        @OptIn(SecretExposure::class)
```

Comment:
this could probably be a file-level opt in in this file

Reply:
as per my other comment. I took care to opt in for every statement individually. Additional context: should something be added or changed wrt. secret exposure: we won't miss it even internally